MalomAPI_cpp/perfect_api.cpp

```diff
@@ -0,0 +1,219 @@
+    s.setStoneCount[W] = Rules::maxKSZ - whiteStonesToPlace;
+    s.setStoneCount[B] = Rules::maxKSZ - blackStonesToPlace;
+    s.kle = onlyStoneTaking;
+    mustBeBetween("playerToMove", playerToMove, 0, 1);
+    s.sideToMove = playerToMove;
+    s.moveCount = 10;
+
+    if (s.futureStoneCount(W) > Rules::maxKSZ) {
+        throw std::invalid_argument("Number of stones in whiteBitboard + "
+                                    "whiteStonesToPlace > " +
+                                    std::to_string(Rules::maxKSZ));
+    }
+    if (s.futureStoneCount(B) > Rules::maxKSZ) {
+        throw std::invalid_argument("Number of stones in blackBitboard + "
+                                    "blackStonesToPlace > " +
+                                    std::to_string(Rules::maxKSZ));
+    }
+
+    std::string errorMsg = s.setOverAndCheckValidSetup();
+    if (errorMsg != "") {
+        throw std::invalid_argument(errorMsg);
+    }
+    if (s.over) {
+        throw std::invalid_argument("Game is already over.");
+    }
+
+    s.lastIrrev = 0;
+
+    int ret = 0;
+
+    try {
+        ret = pp->chooseRandom(pp->goodMoves(s)).toBitBoard();
+    } catch (std::out_of_range &) {
+        throw std::runtime_error("We don't have a database entry for this "
+                                 "position. This can happen either if the "
+                                 "database is corrupted (missing files), or "
+                                 "sometimes when the position is not reachable "
+                                 "from the starting position.");
+    }
+
+    deinitializeIfNeeded();
+
+    return ret;
+}
+
+int MalomSolutionAccess::getBestMoveNoException(
+    int whiteBitboard, int blackBitboard, int whiteStonesToPlace,
+    int blackStonesToPlace, int playerToMove, bool onlyStoneTaking)
+{
+    try {
+        lastError = nullptr;
+        return getBestMove(whiteBitboard, blackBitboard, whiteStonesToPlace,
+                           blackStonesToPlace, playerToMove, onlyStoneTaking);
+    } catch (std::exception &e) {
+        lastError = &e;
+        return 0;
+    }
+}
+
+std::string MalomSolutionAccess::getLastError()
+{
+    if (lastError == nullptr) {
+        return "No error";
+    }
+    return lastError->what();
+}
+
+void MalomSolutionAccess::initializeIfNeeded()
+{
+    if (pp != nullptr) {
+        return;
+    }
+
+    if (sec_val_path == "") {
+        sec_val_path  = ".";
+    }
+
+    Rules::initRules();
+    setVariantStripped();
+    if (!Sectors::hasDatabase()) {
+        throw std::runtime_error("Database files not found in the current "
+                                 "working directory (" +
+                                 std::filesystem::current_path().string() +
+                                 ")");
+    }
+    pp = new PerfectPlayer();
+}
+
+void MalomSolutionAccess::deinitializeIfNeeded()
+{
+    Rules::cleanup();
+
+    if (pp == nullptr) {
+        return;
+    }
```

Comment:
The `Rules::cleanup();` line should be after this `if` statement to avoid double deletions if `deinitializeIfNeeded` is called multiple times. (The "ifNeeded" part suggests that the function is idempotent, i.e., it should be ok to call it redundantly.)

Reply:
Fixed in https://github.com/calcitem/Sanmill/commit/c24e5365b898e157668860ddd6f617333270b09b